lib/ansible/modules/cloud/amazon/cloudfront.py

```diff
@@ -0,0 +1,1765 @@
+                self.module.fail_json(msg="both distribution_id and streaming_distribution_id have been specified. " +
+                                      "please only specify one.")
+            if alias is not None:
+                distribution_id = self.__cloudfront_facts_mgr.get_distribution_id_from_domain_name(
+                    alias)
+            if distribution_id is not None:
+                distribution_response = self.__cloudfront_facts_mgr.get_distribution(
+                    distribution_id)
+            if distribution_response is not None:
+                distribution = distribution_response.get('Distribution')
+                if distribution is not None:
+                    return distribution.get('ARN')
+            streaming_distribution_response = self.__cloudfront_facts_mgr.get_streaming_distribution(
+                streaming_distribution_id)
+            if streaming_distribution_response is not None:
+                streaming_distribution = streaming_distribution_response.get(
+                    'StreamingDistribution')
+                if streaming_distribution is not None:
+                    return streaming_distribution.get('ARN')
+            self.module.fail_json(
+                msg="unable to find a matching distribution with given parameters")
+        except Exception as e:
+            self.module.fail_json(msg="error validating tagging parameters - " + str(e))
+
+    def validate_tags(self, tags):
+        try:
+            if tags is None:
+                return None
+            list_items = []
+            for tag in tags:
+                key = tag.keys()[0]
+                value = str(tag[key])
+                list_items.append({'Key': key, 'Value': value})
+            return list_items
+        except Exception as e:
+            self.module.fail_json(msg="error validating tags - " + str(e))
+
+    def validate_distribution_config_parameters(self, config, default_root_object, is_ipv6_enabled,
+                                                http_version, web_acl_id):
+        try:
+            if default_root_object is not None:
+                config['default_root_object'] = default_root_object
+            else:
+                config['default_root_object'] = ''
+            if is_ipv6_enabled is not None:
+                config['is_i_p_v_6_enabled'] = is_ipv6_enabled
+            else:
+                config['is_i_p_v_6_enabled'] = self.__default_is_ipv6_enabled
+            if http_version is not None:
+                self.valiate_attribute_with_allowed_values(
+                    http_version, 'http_version', self.__valid_http_versions)
+                config['http_version'] = http_version
+            if web_acl_id is not None:
+                config['web_a_c_l_id'] = web_acl_id
+            return config
+        except Exception as e:
+            self.module.fail_json(msg="error validating distribution config parameters - " + s
```

Comment:
I think supporting multiple different resources in one module isn't a sustainable idea. That means we need a test matrix of 4 resource types times all the parameters for all 4. Would it be possible to split these into multiple modules? Looking at the CF API, it looks like we'd want to split:

- "cloudfront_distribution" which would support streaming/non-streaming distributions
- "cloudfront_invalidation" which would be exactly what it sounds like
- "cloudfront_origin_access_identity" which would output the proper S3 canonical user ID. 

Splitting like that would let the cloudfront_distribution module drop at least 8 parameters based on a cursory count.